content/doc/book/pipeline/jenkinsfile.adoc

```
@@ -889,6 +853,124 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
   `withCredentials( ... ) { ... }` step snippet.
 
 
+=== String interpolation
+
+Jenkins Pipeline uses rules identical to link:http://groovy-lang.org[Groovy] for
+string interpolation. Groovy's String interpolation support can be confusing to
+many newcomers to the language. While Groovy supports declaring a string with
+either single quotes, or double quotes, for example:
+
+[source,groovy]
+----
+def singlyQuoted = 'Hello'
+def doublyQuoted = "World"
+----
+
+Only the latter string will support the dollar-sign (`$`) based string
+interpolation, for example:
+
+[source,groovy]
+----
+def username = 'Jenkins'
+echo 'Hello Mr. ${username}'
+echo "I said, Hello Mr. ${username}"
+----
+
+Would result in:
+
+[source]
+----
+Hello Mr. ${username}
+I said, Hello Mr. Jenkins
+----
+
+Understanding how to use string interpolation is vital for using some of
+Pipeline's more advanced features.
+
+[WARNING]
+.Interpolation of sensitive environment variables
+====
+
+Groovy string interpolation should [red]*never* be used with sensitive environment variables (i.e. credentials, see: <<Handling credentials>>).
+This is because the sensitive environment variable will be interpolated during Groovy evaluation, and the environment variable's value could be made available earlier than intended, resulting in sensitive data leaking in various contexts.
+
+For example, consider a sensitive environment variable passed to the `sh` step.
+```
+node {
+  withCredentials([usernamePassword(credentialsId: 'creds', usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')] {
+    /* WRONG! */
+    sh "curl -u ${USERNAME}:${PASSWORD} https://example.com/"
+  }
+}
```

Now that I look at the rest of the page more closely, everything else is using Declarative, so it would probably be best to switch all of your examples to Declarative:

```suggestion
pipeline {
    agent any
    environment {
        EXAMPLE_CREDS = credentials('example-credentials-id')
    }
    stages {
        stage('Example') {
            steps {
                /* WRONG! */
                sh("curl -u ${EXAMPLE_CREDS_USR}:${EXAMPLE_CREDS_PSW} https://example.com/")
            }
        }
    }
}
```

---

swapped to declarative in 881f5b75229a60e43b136a639b47bc5cec2e5e10